### PR TITLE
[Migraphx EP] Static int8 QDQ support

### PR DESCRIPTION
**Pull request by TedThemistokleous**
### Description

Adding static int8 quantization support for MIGraphX Execution Provider

- Allows for parsing in calibration tables generated by Onnxruntime or TensorRT's toolsets
- Add proper environment variables into the MIGraphX EP
- Update python API to include updating execution provider flags -> was missing on python side
- Hook into MIGraphX's int8 quantitation and optimization of models

### Motivation and Context

Required so that we can get onnxruntime to pass in models while leveraging the existing tooling for int8 static QDQ quantization.

First step in a series of PRs which will add further static quantization on the operator level as MIGraphX releases further support. 

These changes drew heavily from the tensorRT EP should allow for similar functionality for GPU based (versus CPU) quantization of models before an inference is performed.

### CODE REVIEW

**Comment by TedThemistokleous** [comment]
@cloudhan @PeixuanZuo 

Keeping this in draft until I can get some models quantized and run through your end to end examples found here: https://github.com/microsoft/onnxruntime-inference-examples/tree/main/quantization

Let me know if you have any further desired pieces I should be adding/taking into consideration. 

I shall be uploading the finished end to end examples to the onnxruntime-inference-example repo as well once completed.

**Comment by TedThemistokleous** [comment]
@cloudhan @PeixuanZuo can you kick off CI for me? We're looking to get support of this out with our release. 

I've confirmed functionality with the tagged end to end example of resnet50.

**Comment by TedThemistokleous** [comment]
@cloudhan can you enable CI again?

**Comment by cloudhan** [comment]
@TedThemistokleous FYI, for quick local development iteration to test the CI, You can grab the CI command from other PRs 
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1173339&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=66420422-c7d6-5f71-625c-4b7851c9b9ba&l=30. Modify it and make it suitable for local test.

**Comment by TedThemistokleous** [comment]
> @TedThemistokleous FYI, for quick local development iteration to test the CI, You can grab the CI command from other PRs https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1173339&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=66420422-c7d6-5f71-625c-4b7851c9b9ba&l=30. Modify it and make it suitable for local test.

Does that also run local linting? That's what seems to be failing here. The training error I'm not seeing. We're currently using the following for builds

./build.sh --config Release  --cmake_extra_defines CMAKE_HIP_COMPILER=/opt/rocm/llvm/bin/clang++ --update --build --build_wheel --parallel --cmake_extra_defines ONNXRUNTIME_VERSION=$(cat ./VERSION_NUMBER) --skip_tests --rocm_home /opt/rocm --use_migraphx --migraphx_home /opt/rocm --rocm_version=`cat /opt/rocm/.info/version-dev` --allow_running_as_root

File: onnxruntime/python/tools/transformers/benchmark.py
Comment (TedThemistokleous) [reply onnxruntime/python/tools/transformers/benchmark.py]: Fixed

Comment (TedThemistokleous) [reply onnxruntime/python/tools/transformers/benchmark.py]: Fixed. Realized this needed to be args.provider

**Comment by TedThemistokleous** [comment]
@cloudhan @ytaous need another kick for CI

**Comment by TedThemistokleous** [comment]
@CloudHan there's nonsensical answers I'm getting from your linter:   

```
810  814 |                           << ". Please reference
    810      |-
https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirements
    811      |-to ensure all dependencies are met.";
         815 |+                             https :  // onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirements
         816 |+to ensure all dependencies are met.";
```

Running lintrunner on onnxruntime root dir gives me the following

```
Warning: Could not find a lintrunner config at: '.lintrunner.private.toml'. Continuing without using configuration file.
WARNING: No previous init data found. If this is the first time you're running lintrunner, you should run `lintrunner init`.
  RUFF failure
  BLACK-ISORT failure
  PYLINT success!
  RUSTFMT success!
  CLANGFORMAT failure


>>> General linter failure:

  Error (RUFF) Linter failed
    Linter failed. This a bug, please file an issue against the linter
    maintainer.
    
    CONTEXT:
    Failed to execute linter command python with args: ["-m",
    "lintrunner_adapters", "run", "ruff_linter", "--config=pyproject.toml",
    "@/tmp/.tmptFQL1X"]
  Error (CLANGFORMAT) Linter failed
    Linter failed. This a bug, please file an issue against the linter
    maintainer.
    
    CONTEXT:
    Failed to execute linter command python with args: ["-m",
    "lintrunner_adapters", "run", "clangformat_linter", "--binary=clang-
    format", "--fallback", "--", "@/tmp/.tmp2KvuTb"]
  Error (BLACK-ISORT) Linter failed
    Linter failed. This a bug, please file an issue against the linter
    maintainer.
    
    CONTEXT:
    Failed to execute linter command python with args: ["-m",
    "lintrunner_adapters", "run", "black_isort_linter", "--", "@/
    tmp/.tmp2PxX1i"]
```

**Comment by cloudhan** [comment]
@TedThemistokleous The ci log is weird because it stripped all whitespace changes from the log, so the log itself make no sense at all. All you need is just run clang-format on that file and it should be automatically formatted. 

For lintrunner, you also need `lintrunner_adapters`. 

`pip install lintrunner lintrunner_adapters` and `lintrunner <the file>` instead of the whole repo will be more helpful.

**Comment by cloudhan** [comment]
For example `lintrunner include/onnxruntime/core/session/onnxruntime_c_api.h`
image



`lintrunner -a include/onnxruntime/core/session/onnxruntime_c_api.h` will automatically fix it for you:


```console
> git diff                                                                                                                                                                                               (base) 
diff --git a/include/onnxruntime/core/session/onnxruntime_c_api.h b/include/onnxruntime/core/session/onnxruntime_c_api.h
index c44b941c13..8f22f2f75e 100644
--- a/include/onnxruntime/core/session/onnxruntime_c_api.h
+++ b/include/onnxruntime/core/session/onnxruntime_c_api.h
@@ -598,12 +598,11 @@ typedef struct OrtTensorRTProviderOptions {
  * \see OrtApi::SessionOptionsAppendExecutionProvider_MIGraphX
  */
 typedef struct OrtMIGraphXProviderOptions {
-
-  int device_id;             // hip device id.
-  int migraphx_fp16_enable;  // enable MIGraphX FP16 precision. Default 0 = false, nonzero = true
-  int migraphx_int8_enable;  // enable MIGraphX INT8 precision. Default 0 = false, nonzero = true
-  int migraphx_use_native_calibration_table;          // MIGraphx INT8 cal table. Default 0 = false, noznero = true
-  const char* migraphx_int8_calibration_table_name;   // MIGraphx INT8 calibration table name
+  int device_id;                                     // hip device id.
+  int migraphx_fp16_enable;                          // enable MIGraphX FP16 precision. Default 0 = false, nonzero = true
+  int migraphx_int8_enable;                          // enable MIGraphX INT8 precision. Default 0 = false, nonzero = true
+  int migraphx_use_native_calibration_table;         // MIGraphx INT8 cal table. Default 0 = false, noznero = true
+  const char* migraphx_int8_calibration_table_name;  // MIGraphx INT8 calibration table name
 } OrtMIGraphXProviderOptions;
 
 /** \brief OpenVINO Provider Options
```

**Comment by TedThemistokleous** [comment]
> For example `lintrunner include/onnxruntime/core/session/onnxruntime_c_api.h` image
> 
> `lintrunner -a include/onnxruntime/core/session/onnxruntime_c_api.h` will automatically fix it for you:
> 
> ```
> > git diff                                                                                                                                                                                               (base) 
> diff --git a/include/onnxruntime/core/session/onnxruntime_c_api.h b/include/onnxruntime/core/session/onnxruntime_c_api.h
> index c44b941c13..8f22f2f75e 100644
> --- a/include/onnxruntime/core/session/onnxruntime_c_api.h
> +++ b/include/onnxruntime/core/session/onnxruntime_c_api.h
> @@ -598,12 +598,11 @@ typedef struct OrtTensorRTProviderOptions {
>   * \see OrtApi::SessionOptionsAppendExecutionProvider_MIGraphX
>   */
>  typedef struct OrtMIGraphXProviderOptions {
> -
> -  int device_id;             // hip device id.
> -  int migraphx_fp16_enable;  // enable MIGraphX FP16 precision. Default 0 = false, nonzero = true
> -  int migraphx_int8_enable;  // enable MIGraphX INT8 precision. Default 0 = false, nonzero = true
> -  int migraphx_use_native_calibration_table;          // MIGraphx INT8 cal table. Default 0 = false, noznero = true
> -  const char* migraphx_int8_calibration_table_name;   // MIGraphx INT8 calibration table name
> +  int device_id;                                     // hip device id.
> +  int migraphx_fp16_enable;                          // enable MIGraphX FP16 precision. Default 0 = false, nonzero = true
> +  int migraphx_int8_enable;                          // enable MIGraphX INT8 precision. Default 0 = false, nonzero = true
> +  int migraphx_use_native_calibration_table;         // MIGraphx INT8 cal table. Default 0 = false, noznero = true
> +  const char* migraphx_int8_calibration_table_name;  // MIGraphx INT8 calibration table name
>  } OrtMIGraphXProviderOptions;
>  
>  /** \brief OpenVINO Provider Options
> ```

Looks like there's an assumption here with python versions. I got it sorted and reran everything

**Comment by cloudhan** [comment]
@justinchuby Is it possible to set the Lint pipeline to automatically run without needing member approve?

**Comment by justinchuby** [comment]
> @justinchuby Is it possible to set the Lint pipeline to automatically run without needing member approve?

I think that would be a repo policy the team has to set. I think only new contributors need approval and as far as I know it is a repo level settings. 

So that may be hard to set it just on lint

**Comment by cloudhan** [comment]
@justinchuby 
> only new contributors need approval

Not true anymore, now every commit needs an approval...

**Comment by justinchuby** [comment]
> @justinchuby
> 
> > only new contributors need approval
> 
> Not true anymore, now every commit needs an approval...

I mean once the first PR (this) is merged, one will be a known contributor and no longer required approval I think?

**Comment by cloudhan** [comment]
@justingchuby Ted is already a contributor:
image

Is possible that we add some CODEONWERS so that like AMD can contribute to ROCm and MIGraphX, without our intervention, except for the final merge action. Likewise, for Nvidia, they should also have representative for CUDA and TensorRT.

**Comment by justinchuby** [comment]
I see. Maybe @snnn or @faxu can help with configuring the repository policy? Thanks!

**Comment by TedThemistokleous** [comment]
> @justinchuby Ted is already a contributor: image
> 
> Is possible that we add some CODEONWERS so that like AMD can contribute to ROCm and MIGraphX, without our intervention, except for the final merge action. Likewise, for Nvidia, they should also have representative for CUDA and TensorRT.

This would speed things up for us greatly. I tent to ping/pull in a few people when I've committed changes.  Let me know what's needed on our end to be able to trigger jobs.

**Comment by TedThemistokleous** [comment]
> I see. Maybe @snnn or @faxu can help with configuring the repository policy? Thanks!

@snnn or @faxu , let me know if this is possible. I have no way of working against your CI without pinging people on your team constantly.

**Comment by faxu** [comment]
Unfortunately, our policy requires Microsoft employees to kick off CI pipelines in this repo.

**Comment by TedThemistokleous** [comment]
> Unfortunately, our policy requires Microsoft employees to kick off CI pipelines in this repo.

Sure, understood. So the other pipelines will be held up then? 

I'm seeing oddities in your logs like

```
  File "C:\vss-agent\2.311.0\_work\onnxruntime\onnxruntime\tools\python\util\run.py", line 49, in run
    completed_process = subprocess.run(
                        ^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.6\x64\Lib\subprocess.py", line 571, in run
Failed to import psutil. Please `pip install psutil` for better estimation of nvcc threads. Use nvcc_threads=1
    raise CalledProcessError(retcode, process.args,
```

Which has nothing to do with my changes but looks like a "fail" in CI. 

@cloudhan @PeixuanZuo @ytaous

**Comment by faxu** [comment]
/azp run onnxruntime-python-checks-ci-pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, 
ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline

**Comment by TedThemistokleous** [comment]
Hi @faxu @cloudhan is there anything else you'd like to see with this PR?
 Its been two weeks and I haven't gotten any feedback/reviewers.

**Comment by faxu** [comment]
/azp run onnxruntime-python-checks-ci-pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, 
ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline

**Comment by faxu** [comment]
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline,
Linux OpenVINO CI Pipeline, ONNX Runtime Web CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, Linux QNN CI Pipeline, Linux QNN CI Pipeline, Windows x64 QNN CI Pipeline

**Comment by faxu** [comment]
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline,
Linux OpenVINO CI Pipeline, ONNX Runtime Web CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, orttraining-linux-ci-pipeline

**Comment by TedThemistokleous** [comment]
Can anyone give me a rerun? @faxu @cloudhan

**Comment by jeffdaily-ms** [comment]
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline,
Linux OpenVINO CI Pipeline, ONNX Runtime Web CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, orttraining-linux-ci-pipeline

**Comment by jeffdaily-ms** [comment]
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline,
Linux OpenVINO CI Pipeline, ONNX Runtime Web CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, orttraining-linux-ci-pipeline

**Comment by TedThemistokleous** [comment]
@cloudhan / @faxu let me know if there's anything I need for this. 

Currently also seeing issues off your mainline with Eigen 3.4 not pulling down anymore as part of the build process out side of this change (onnxruntime:main) 

is there another fix/commit I should be pulling into this/rebasing off of?

**Comment by TedThemistokleous** [comment]
@cloudhan /@faxu rebased changes off master since the eigen issue was breaking local builds.

**Comment by faxu** [comment]
/azp run onnxruntime-python-checks-ci-pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, 
ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, Linux QNN CI Pipeline, Linux QNN CI Pipeline, Windows x64 QNN CI Pipeline

**Comment by PeixuanZuo** [comment]
@TedThemistokleous , could you please take a look about failed lint c++ CI?

**Comment by jeffdaily** [comment]
> @TedThemistokleous , could you please take a look about failed lint c++ CI?

@PeixuanZuo there is something strange about linting.  For some reason our ROCm-related PRs flag far more linting issues than other non-ROCm PRs.  It really hampers how quickly we can land PRs.  For example, it flags include/onnxruntime/core/session/onnxruntime_c_api.h:603 which is inside the struct OrtMIGraphXProviderOptions, but looking 10 lines above inside the struct OrtTensorRTProviderOptions there are 3 lines over 120 long.  It's only flagging our newly added lines.  Are we being held to a higher standard than other EPs?

**Review by jeffdaily-ms — APPROVED** [review]

**Comment by PeixuanZuo** [comment]
/azp run onnxruntime-python-checks-ci-pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed,
ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, Linux QNN CI Pipeline, Linux QNN CI Pipeline, Windows x64 QNN CI Pipeline, Windows ARM64 QNN CI Pipeline, Linux MIGraphX CI Pipeline

**Review by PeixuanZuo — APPROVED** [review]

**Pull request merged**
